### PR TITLE
fix: cast tokenizer & lam to full precision before loss calculation

### PR DESCRIPTION
**Pull request by emergenz**

### CODE REVIEW

File: train_lam.py [hunk 1]
```diff
@@ -83,6 +83,7 @@ def lam_loss_fn(
     inputs["videos"] = inputs["videos"].astype(args.dtype) / 255.0
     model.train()
     outputs = model(inputs, training=True)
+    outputs["recon"] = outputs["recon"].astype(jnp.float32)
     gt_future_frames = inputs["videos"][:, 1:]
```
Comment (maharajamihir) [inline train_lam.py:87]: in this case gt is not in full precision though. If you're doing this, it likely makes sense to cast the gt ones to full precision as well?

**Review by maharajamihir — CHANGES_REQUESTED** [review]
in this case gt is not in full precision though. If you're doing this, it likely makes sense to cast the gt ones to full precision as well?

**Comment by maharajamihir** [comment]
> Ah lol good catch, if gt is in int8, then we should cast all reconstructions to int8 before doing the loss calculation, right?

a few lines above GT is casted to bfloat16 and normalized. I think casting it to float32 is sufficient. 
Conversion to int8 would result in quantisation of the values. imo, loss calc should be in float32 for highest granularity.
@emergenz

File: train_dynamics.py [hunk 2]
```diff
@@ -94,19 +94,21 @@ def dynamics_loss_fn(
...
 ) -> tuple[jax.Array, tuple[jax.Array, dict]]:
     """Compute masked dynamics loss"""
-    inputs["videos"] = inputs["videos"].astype(args.dtype) / 255.0
+    gt = jnp.asarray(inputs["videos"], dtype=jnp.float32) / 255.0
+    inputs["videos"] = gt.astype(args.dtype)
     model.train()
     outputs = model(inputs, training=True)
     mask = outputs["mask"]
     outputs["token_logits"] = outputs["token_logits"].astype(jnp.float32)
+    outputs["video_tokens"] = outputs["video_tokens"].astype(jnp.int32)
```
[the remark this answers is not in the document]
Comment (emergenz) [reply train_dynamics.py:103]: Casting the 'gt' to full precision before the loss calculation.
